README.md
# bibliometRics [![DOI](https://zenodo.org/badge/57246514.svg)](https://zenodo.org/badge/latestdoi/57246514)
Santiago Beguería  
26 April 2016
    
`bibliometRics` is an R package for bibliometric analysis of scientific production.
It can be used for analysing the production of a single author, a working team, department, institute, etcetera.

This document describes the main functionalities in the package, and how to do a bibliometric analysis with `bibliometRics`, including producing automatic pdf reports via `knitr`.

You can cite this package on your work as: **Beguería S. (2015) bibliometRics: an R package for for bibliometric analysis of scientific production, doi:10.5281/zenodo.834260**.

## Installation

First of all, make sure you installed the package by sourcing it (note that this is a working project, so no 'oficial' package has been created yet).

```r
source('bibliometRics.R')
```


## Getting some bibliometric data

So far the unique source of blibiometric information accepted by `bibliometRics` is the Web of Science (WoS) by Thomson Reuters, but other sources such as Scopus or Google Scholar can be added in the future.
Publications can be selected for a given author (easiest if you know its author ID) or a group (such a research group or a department, for instance).
The WoS allows to generate a citation report, stating the number of citations received by each item, every year after its publication.
So once you are good with the selection of publications you want to analyze, you can click on the 'create citation report' button and then select the 'save to text file' option.
This will generate a text file and download it to your computer.
You may want to edit the AUTHOR field on the first line of this text file, although this is not strictly necessary for doing the analysis.

An example citation report file from the WoS is the file [sbegueria.txt](sbegueria.txt), which contains data of papers I have authored, as of January 2017.

The core function for reading WoS citation report data is, not surprisingly, `read.wos`. 
Its only argument is the name of the data file you want to read:


```r
bib <- read.wos('sbegueria.txt')
str(bib)
```

```r
## List of 3
##  $ author   : chr "BEGUERIA,  SANTIAGO"
##  $ reference: chr "sbegueria"
##  $ pubs     :'data.frame':	88 obs. of  39 variables:
##   ..$ Title            : chr [1:88] "Estimating erosion rates using Cs-137 measurements and WATEM/SEDEM in a Mediterranean cultivated field" "Recent changes and drivers of the atmospheric evaporative demand in the Canary Islands" "Mid and late Holocene forest fires and deforestation in the subalpine belt of the Iberian range, northern Spain" "Use of disdrometer data to evaluate the relationship of rainfall kinetic energy and intensity (KE-I)" ...
##   ..$ Authors          : chr [1:88] "Quijano, Laura; Begueria, Santiago; Gaspar, Leticia; Navas, Ana" "Vicente-Serrano, Sergio M.; Azorin-Molina, Cesar; Sanchez-Lorenzo, Arturo; El Kenawy, Ahmed; Martin-Hernandez, Natalia; Pena-Ga"| __truncated__ "Garcia-Ruiz, Jose M.; Sanjuan, Yasmina; Gil-Romera, Graciela; Gonzalez-Samperiz, Penelope; Begueria, Santiago; Arnaez, Jose; Co"| __truncated__ "Angulo-Martinez, M.; Begueria, S.; Kysely, J." ...
##   ..$ Corporate Authors: logi [1:88] NA NA NA NA NA NA ...
##   ..$ Editors          : logi [1:88] NA NA NA NA NA NA ...
##   ..$ Book Editors     : chr [1:88] "" "" "" "" ...
##   ..$ Source Title     : chr [1:88] "CATENA" "HYDROLOGY AND EARTH SYSTEM SCIENCES" "JOURNAL OF MOUNTAIN SCIENCE" "SCIENCE OF THE TOTAL ENVIRONMENT" ...
##   ..$ Publication Date : chr [1:88] "MAR 2016" "AUG 23 2016" "OCT 2016" "OCT 15 2016" ...
##   ..$ Publication Year : int [1:88] 2016 2016 2016 2016 2016 2016 2016 2015 2015 2015 ...
##   ..$ Volume           : chr [1:88] "138" "20" "13" "568" ...
##   ..$ Issue            : chr [1:88] "" "8" "10" "" ...
##   ..$ Part Number      : chr [1:88] "" "" "" "" ...
##   ..$ Supplement       : logi [1:88] NA NA NA NA NA NA ...
##   ..$ Special Issue    : chr [1:88] "" "" "" "" ...
##   ..$ Beginning Page   : int [1:88] 38 3393 1760 83 2120 3413 NA 429 853 773 ...
##   ..$ Ending Page      : chr [1:88] "51" "3410" "1772" "94" ...
##   ..$ Article Number   : chr [1:88] "" "" "" "" ...
##   ..$ DOI              : chr [1:88] "10.1016/j.catena.2015.11.009" "10.5194/hess-20-3393-2016" "10.1007/s11629-015-3763-8" "10.1016/j.scitotenv.2016.05.223" ...
##   ..$ Conference Title : chr [1:88] "" "" "" "" ...
##   ..$ Conference Date  : chr [1:88] "" "" "" "" ...
##   ..$ Total Citations  : int [1:88] 0 0 0 0 2 2 4 0 4 4 ...
##   ..$ Average per Year : num [1:88] 0 0 0 0 1 1 2 0 1.33 1.33 ...
##   ..$ 2000             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2001             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2002             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2003             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2004             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2005             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2006             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2007             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2008             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2009             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2010             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2011             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2012             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2013             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2014             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ 2015             : int [1:88] 0 0 0 0 0 0 0 0 0 1 ...
##   ..$ 2016             : int [1:88] 0 0 0 0 2 2 4 0 4 3 ...
##   ..$ 2017             : int [1:88] 0 0 0 0 0 0 0 0 0 0 ...
```

The result is a list with the following three elements:
* `author`, a character object with the name of the author being analysed;
* `reference`, a character object; and
* `pubs`, a data frame with the publications in rows, and the data referring to each publications in columns, including the number of citations received, year by year.


## Analyzing bibliometric data

The main function for analyizing these data is `bibliometric`.
It takes an object with the bibliometric data resulting from a call to `read.wos` and returns a data.frame with a number of bibliometric indices.

In order to compute some metrics which are based on the number of citations received by each publication, data on the percentile baselines for each scientific discipline is requireed.
This data can be found on another Thomson Reuters product, the 'Essential Science Indicators' database.
You need to navigate to the 'Field Baselines' tab, and then select 'Percentiles' to get the field percentile baselines (FPB).
You can then download the FPB table in csv format.

An example FPB table as of January 2017 can be found in file [BaselinePercentiles.csv](BaselinePercentiles.csv).
This table can be read with the function 'read.baselines':

```r
base <- read.baselines('BaselinePercentiles.csv')
```

The resulting object is a list with as many items as disciplines.
For each discipline, a data.frame is stored containing the threshold number of citations corresponding to each percentile, depending on the publication year.
For instance, this is are the baseline thresholds for the discipline 'Geosciences':

```r
base$GEOSCIENCES
```

We can now use the function `bibliometric`, especifying the, the field baselines, and the discipline (the table for 'all fields' will be used as a default if no discipline is specified).

```r
bibliometric(bib, base, 'GEOSCIENCES')
```

```r
##                   name  ini years pubs lead pubs_year hin hin_year
##  1 BEGUERIA,  SANTIAGO 2000    16   88   17       5.5  33     2.06
##    gin gin_year cit_tot cit_year cit_art ifact2 ifact5 i10 i25 i50
##  1  57     3.56    3472      217   39.45  15.43  14.13  65  48  16
##    cit_max pubs09 pubs09_lead pubs099 iscore iscore_lead
##  1     616     32           7      11   4145         187
```

These are the following:

 Label         | Meaning
---------------| ----------------------------------------------------
 name          | Name of the author, group, department, etc.
 ini           | Initial year of the publication record
 span          | Time span (years) of the publication record
 pubs          | Total number of publications
 lead          | Total number of publications, as the lead (first) author
 pubs_year     | Mean number of publications per year
 hin           | Hirsch's h-index
 hin_year      | h-index per year
 gin           | Egghe's g-index
 gin_year      | g-index per year
 cit_tot       | Total number of citations
 cit_year      | Mean number of citations per year 
 cit_art       | Mean number of citations per article
 ifact2        | Impact factor, computed over the last two years
 ifact5        | Impact factor, computed over the last five years
 i10           | Number of publications with 10 or more citations
 i25           | Number of publications with 25 or more citations
 i50           | Number of publications with 50 or more citations
 cit_max       | Number of citations of the most citated publications
 pubs09        | Number of publications over the 90th percentile in its discipline
 pubs09_lead   | Number of publications over the 90th percentile in its discipline, as lead author
 pubs099       | Number of publications over the 99th percentile in its discipline
 iscore        | i-score
 iscore_lead   | i-score, as lead author

There are functions for computing some of the indices, such as the Hirsch and the Egghe indices.
These can be computed for the whole period analyized, or up to a given year.


```r
hirsch(bib)
```

```r
##  [1] 33
```

```r
egghe(bib)
```

```r
##  [1] 57
```

```r
hirsch(bib, 2010)
```

```r
##  [1] 13
```

```r
egghe(bib, 2010)
```

```r
##  [1] 45
```

There is also a function for ranking the publications in quantiles:

```r
rank(bib, q=base$GEOSCIENCES)
```

```r
##   [1] >q0     >q0     >q0     >q0     >q0.9   >q0.9   >q0.99  >q0    
##   [9] >q0.8   >q0.8   >q0.9   >q0.9   >q0.99  >q0.99  >q0.99  >q0.999
##  [17] >q0     >q0.99  >q0.99  >q0.5   >q0.9   >q0.9   >q0.999 >q0    
##  [25] >q0     >q0.5   >q0.5   >q0.8   >q0.8   >q0.9   >q0.9   >q0.9  
##  [33] >q0     >q0     >q0.5   >q0.5   >q0.8   >q0.8   >q0.8   >q0.8  
##  [41] >q0.9   >q0.9   >q0.99  >q0     >q0     >q0.5   >q0.8   >q0.9  
##  [49] >q0.9   >q0.9   >q0.9   >q0.9   >q0.99  >q0.999 >q0.5   >q0.5  
##  [57] >q0.8   >q0.8   >q0.8   >q0.8   >q0.8   >q0.9   >q0.9   >q0    
##  [65] >q0.5   >q0.8   >q0.8   >q0.8   >q0.8   >q0.5   >q0.5   >q0.8  
##  [73] >q0.8   >q0.9   >q0.9   >q0.9   <NA>    <NA>    <NA>    <NA>   
##  [81] <NA>    <NA>    <NA>    <NA>    <NA>    <NA>    <NA>    <NA>   
##  Levels: >q0.9999 >q0.999 >q0.99 >q0.9 >q0.8 >q0.5 >q0
```

```r
table(rank(bib, q=base$GEOSCIENCES))
```
```r
##  >q0.9999  >q0.999   >q0.99    >q0.9    >q0.8    >q0.5      >q0 
##         0        3        8       21       20       11       13 
```


A specific plotting function makes it easy to resume most of this information in graphic form.
The plots also inform about the temporal evolution of the bibliometric indices, which may be useful for evaluating the scientific career of the evaluated.


```r
biblioplot(bib, q=base$GEOSCIENCES)
```

<div class="figure" style="text-align: center">
<img src="./biblioplot_example.pdf" alt="A graphic bibliometric analysis" />
<p class="caption">A graphic bibliometric analysis</p>
</div>

The first plot reflects the productivity (quantity) of the author, as well as its impact (citations received). It shows the cumulative number of publications, with distinction between the publications as lead author (black bars) and those as co-author (white bars). The plot also showcases the number of citations for all the publications (white circles) and for those as lead author (black circles). There is a fixed ratio of 1:10 between the left (publications) and the right (citations) axis, allowing for easy comparison across authors, groups, etc. This ratio implies assuming a mean citation rate of 10 citations per article publishes (a rate that is, of course, arbitrary).

The second plot focuses on the impact of the publications. It shows the annual evolution of the Hirsch's h-index (white circles) and the Egghe's g-index (black circles), with a fixed ratio of 1:2 between them. Evolution of the h-index is compared with an 1:1 evolution (dashed line), since it is usually assumed that the h-index grows, as an average, at a rate of 1 per year.

The third plot attempts at evaluating the excelence of the publications. It shows the number of publications classified by quantiles, according to the ISI-WoK Scientific Indicators per discipline. For each quantile, the total number of publications is shown (white bars), as well as the publications as lead author (black bars).

A formatted list of all the publications is also produced.

```r
format_pub(cbind(bib$pubs, rank(bib,q=base$GEOSCIENCES))[1,], au=bib$au)
```

```r
##  [1] "\\item Quijano, Laura;\\textbf{ Begueria, Santiago}; Gaspar, Leticia; Navas, Ana. Estimating Erosion Rates using Cs-137 Measurements and Watem/Sedem in a Mediterranean Cultivated Field. \\textit{CATENA} 138: 38--51. 2016. (cit: 0; $7$)\n"
```


## Automated bibliometric reports

The package also contains a template `bibliometRics.Rtex` file, useful for creating automated reports.
You'll need to load the package `knitr` in order to produce the report.

```r
require(knitr)
```


```r
infile <- 'sbegueria.txt'
outfile <- 'sbegueria.Rtex'

# Create custom .Rtex file from the template and knit it
x <- readLines('bibliometRics.Rtex')
x <- gsub('FILENAME',infile,x)
write(x,outfile)
knit(outfile)

# Compile the resulting .tex file and create a .pdf from it
system(paste('/Library/TeX/texbin/pdflatex ',
	gsub('.txt','',infile),'.tex',sep=''))

# Remove unnecesary intermediate files
kk <- list.files('.',paste(gsub('.txt','', infile)))
file.remove(kk[-c(grep('.pdf',kk),grep('.txt',kk))])
```

An example report generated by the above code chunk can be found in the file [sbegueria.pdf]('./sbegueria.pdf').


NEWS.md
# blah 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(bibliometRics)

test_check("bibliometRics")


tests/testthat/test-biblio.R
testthat::context("Just a simple test")


testthat::test_that("bibliometric works", {

  wos_file = system.file("sbegueria.txt", package = "bibliometRics")
  testthat::expect_silent({
    bib <- read.wos(wos_file)
  })
  base = read.baselines()
  discipline = 'GEOSCIENCES'
  testthat::expect_silent({
    bibliometric(bib, base, 'GEOSCIENCES')
  })
})


R/plot.R
#' Plot Bibliometrics
#'
#' @param bib A list with an element called  \code{pubs}, which is a
#' \code{data.frame} with columns \code{'Publication Year'} and
#' \code{'Total Citations'}.  Usually
#' from \code{\link{read.wos}}
#' @param quant Quantile, passed to \code{\link{citation_rank}}
#'
#' @return NULL
#' @export
#'
#' @import graphics
#'
#' @examples
#' wos_file = system.file("sbegueria.txt", package = "bibliometRics")
#' bib <- read.wos(wos_file)
#' base = read.baselines()
#' biblioplot(bib, q=base$GEOSCIENCES)
#' bib$author = gsub('Ñ', 'N', bib$author)
biblioplot <- function(bib, quant) {
  opar = par(no.readonly = TRUE)
  on.exit({
    par(opar)
  })
  par(mfrow = c(1, 3))

  au <- bib$author
  dat <- bib$pubs

  # Publication range
  pubrange <- range(dat[, 'Publication Year'])

  # Leading authors
  leads <-
    as.character(lapply(dat[, 'Authors'], function(x) {
      strsplit(x, ';')[[1]][1]
    }))
  w_lead <-
    grep(strsplit(au, ',')[[1]][1], leads, ignore.case = TRUE)

  # Publications and citations
  # pubs per year
  years <- dat[, 'Publication Year']
  years <- factor(years, levels = as.character(min(years):max(years)))
  years_nolead <- years[-w_lead]
  years_lead <- years[w_lead]
  pubs <- rbind(table(years_lead), table(years_nolead))
  rownames(pubs) <- c('pubs_lead', 'pubs_nolead')
  # cumulative pubs
  pubs_cum <- t(apply(pubs, 1, cumsum))
  # citations per year
  cits_lead <-
    t(as.matrix(colSums(bib$pubs[w_lead, 22:ncol(bib$pubs)])))
  cits_nolead <-
    t(as.matrix(colSums(bib$pubs[-w_lead, 22:ncol(bib$pubs)])))
  cits <- rbind(cits_lead, cits_nolead)
  rownames(cits) <- c('cits_lead', 'cits_nolead')
  w <- colnames(cits) >= pubrange[1] & colnames(cits) <= pubrange[2]
  cits <- cits[, w]
  # cumulative citations
  cits_cum <- t(apply(cits, 1, cumsum))
  # plot
  par(mar = c(5, 4, 4, 4) + 0.1)
  ylabs <-
    pretty(c(0, max(max(
      colSums(pubs_cum)
    ), ceiling(max(
      colSums(cits_cum)
    ) / 10))))
  ylims <- c(min(ylabs), max(ylabs))
  pubbar <- barplot(pubs_cum,
                    ylim = ylims,
                    ylab = '',
                    yaxt = 'n')
  abline(h = ylabs, col = 'lightgray', lty = 'dotted')
  par(new = TRUE)
  pubbar <- barplot(pubs_cum,
                    ylim = ylims,
                    ylab = '',
                    yaxt = 'n')
  lines(x = pubbar, y = cits_cum['cits_lead', ] / 10)
  points(x = pubbar, y = cits_cum['cits_lead', ] / 10, pch = 19)
  lines(x = pubbar, y = colSums(cits_cum) / 10)
  points(
    x = pubbar,
    y = colSums(cits_cum) / 10,
    pch = 19,
    col = 'lightgray'
  )
  points(x = pubbar,
         y = colSums(cits_cum) / 10,
         pch = 21)
  axis(2)
  axis(4, at = pretty(ylims), labels = pretty(ylims) * 10)
  mtext('Pubs', 2, line = 2.5, cex = 0.75)
  mtext('Citations', 4, line = 2.5, cex = 0.75)

  # h- and g-index
  years <-
    c(min(dat[, 'Publication Year']):max(dat[, 'Publication Year']))
  hh <- NULL
  gg <- NULL
  for (y in years) {
    hh <- c(hh, hirsch(bib, y))
    gg <- c(gg, egghe(bib, y))
  }
  # plot
  ylabs <- pretty(c(0, max(max(hh), ceiling(max(
    gg
  ) / 2))))
  ylims <- c(min(ylabs), max(ylabs))
  plot(
    x = years,
    y = hh,
    type = 'n',
    ylab = '',
    xlab = '',
    ylim = ylims
  )
  abline(h = pretty(hh), col = 'lightgray', lty = 'dotted')
  lines(years, 0:{
    length(years) - 1
  }, lty = 'dashed')
  lines(x = years, y = hh)
  points(x = years, y = hh, pch = 19)
  mtext('h-index', 2, line = 2.5, cex = 0.75)
  par(new = TRUE)
  lines(x = years, y = gg / 2)
  points(
    x = years,
    y = gg / 2,
    pch = 19,
    col = 'lightgray'
  )
  points(x = years, y = gg / 2, pch = 21)
  axis(4, at = pretty(ylims), labels = pretty(ylims) * 2)
  mtext('g-index', 4, line = 2.5, cex = 0.75)

  # Excelence
  exc <-
    rbind(rev(table(
      citation_rank(bib, w_lead, quant = quant)
    )), rev(table(
      citation_rank(bib,-w_lead, quant = quant)
    )))[, -7]
  rownames(exc) <- c('pubs_lead', 'pubs_nolead')
  # plot
  barplot(exc)
  abline(h = pretty(colSums(exc)),
         col = 'lightgray',
         lty = 'dotted')
  par(new = TRUE)
  barplot(exc)
  mtext('Counts', 2, line = 2.5, cex = 0.75)
}
#biblioplot(bib)

#' @rdname biblioplot
#' @export
biblioplot2 <- function(bib) {
  opar = par(no.readonly = TRUE)
  on.exit({
    par(opar)
  })
  par(mfrow = c(1, 2))

  au <- bib$author
  dat <- bib$pubs

  # Publication range
  pubrange <- range(dat[, 'Publication Year'])

  # Leading authors
  leads <-
    as.character(lapply(dat[, 'Authors'], function(x) {
      strsplit(x, ';')[[1]][1]
    }))

  # Publications and citations
  # pubs per year
  years <- dat[, 'Publication Year']
  years <- factor(years, levels = as.character(min(years):max(years)))
  pubs <- table(years)
  # cumulative pubs
  pubs_cum <- cumsum(pubs)
  # citations per year
  cits <- t(as.matrix(colSums(bib$pubs[, 22:ncol(bib$pubs)])))
  rownames(cits) <- c('cits')
  w <- colnames(cits) >= pubrange[1] & colnames(cits) <= pubrange[2]
  cits <- cits[, w]
  # cumulative citations
  cits_cum <- cumsum(cits)
  # plot
  par(mar = c(5, 4, 4, 4) + 0.1)
  ylabs <- pretty(c(0, max(max(pubs_cum), ceiling(max(
    cits_cum
  ) / 10))))
  ylims <- c(min(ylabs), max(ylabs))
  pubbar <- barplot(pubs_cum,
                    ylim = ylims,
                    ylab = '',
                    yaxt = 'n')
  abline(h = ylabs, col = 'lightgray', lty = 'dotted')
  par(new = TRUE)
  pubbar <- barplot(pubs_cum,
                    ylim = ylims,
                    ylab = '',
                    yaxt = 'n')
  lines(x = pubbar, y = cits_cum / 10)
  points(x = pubbar, y = cits_cum / 10, pch = 19)
  axis(2)
  axis(4, at = pretty(ylims), labels = pretty(ylims) * 10)
  mtext('Pubs', 2, line = 2.5, cex = 0.75)
  mtext('Citations', 4, line = 2.5, cex = 0.75)
  title('Publications and citations')

  # h-index
  years <-
    c(min(dat[, 'Publication Year']):max(dat[, 'Publication Year']))
  hh <- NULL
  for (y in years) {
    hh <- c(hh, hirsch(bib, y))
    gg <- c(gg, egghe(bib, y))
  }
  # plot
  ylabs <- pretty(c(0, max(max(hh))))
  ylims <- c(min(ylabs), max(ylabs))
  plot(
    x = years,
    y = hh,
    type = 'n',
    ylab = '',
    xlab = '',
    ylim = ylims
  )
  abline(h = pretty(hh), col = 'lightgray', lty = 'dotted')
  lines(years, 0:{
    length(years) - 1
  }, lty = 'dashed')
  lines(x = years, y = hh)
  points(x = years, y = hh, pch = 19)
  mtext('h-index', 2, line = 2.5, cex = 0.75)
  title('h-index')
  return(invisible(NULL))
}





R/hirsch.R

#' Calculate Hirsch's h-index or Egghe's g-index
#'
#' @param bib A list with an element called  \code{pubs}, which is a
#' \code{data.frame} with column \code{'Publication Year'}.  Usually
#' from \code{\link{read.wos}}
#' @param year Maximum year to include for calculation
#'
#' @return A scalar number
#' @export
#'
#' @examples
#' wos_file = system.file("sbegueria.txt", package = "bibliometRics")
#' bib <- read.wos(wos_file)
#' hirsch(bib)
#' hirsch(bib, 2010)
#' egghe(bib)
#' egghe(bib, 2010)
#' ifactor(bib) # 2-year impact factor
#' ifactor(bib,2013) # 2-year impact factor, year 2013
#' ifactor(bib,n=5)  # 5-year impact factor#'
hirsch <- function(bib, year = NULL) {
  x <- bib$pubs
  if (is.null(year)) {
    #y <- max(x[,'Publication Year'])
    year <- rev(colnames(x))[1]
  }
  w <- which(x[, 'Publication Year'] <= year)
  ww <- which(colnames(x) <= year)
  if (length(ww) > 1) {
    cit <- rowSums(x[w, ww])
  } else {
    cit <- x[w, ww]
  }
  #cit <- x[,'Total Citations']
  hin <- cbind(cit[order(cit, decreasing = TRUE)], 1:length(cit))
  w <- (hin[, 1] - hin[, 2]) > 0
  if (sum(w) == 0)
    return(0)
  return(max(hin[w, 2]))
}


#' @export
#' @rdname hirsch
egghe <- function(bib, year = NULL) {
  x <- bib$pubs
  if (is.null(year)) {
    year <- max(x[, 'Publication Year'])
  }
  w <- x[, 'Publication Year'] <= year
  cit <- x[w, 'Total Citations']
  gin <- cbind(1:length(cit),
               cumsum(cit[order(cit, decreasing = TRUE)]),
               (1:length(cit)) ^ 2)
  w <- (gin[, 2] - gin[, 3]) > 0
  return(max(gin[w, 1]))
}


#' @export
#' @rdname hirsch
#' @param n number of years
ifactor <- function(bib, year = NULL, n = 2) {
  x <- bib$pubs
  all_years <- x[, 'Publication Year']
  if (is.null(year)) {
    year <- max(all_years, na.rm = TRUE) - 1
  }
  w <- which(all_years >= (year - n) & all_years < year)
  ww <- which(colnames(x) == year)
  sum(x[w, ww]) / length(w)
}


R/read.wos.R
# Functions to read and process Web of Science citation report data.
#bib <- read.isiwok('isiwok_MAM.txt')

#' Read and process Web of Science citation report data.
#'
#' @param infile Input text file from Web of Science
#'
#' @return A list of information
#' @export
#' @examples
#' wos_file = system.file('sbegueria.txt', package = "bibliometRics")
#' bib <- read.wos(wos_file)
read.wos <- function(infile) {

  # author
  au <- read.table(infile, sep=',', nrows=1, stringsAsFactors=FALSE)
  #au <- paste(gsub('AUTHOR: \\(','',au[1]),gsub('\\)','',au[2]),sep=', ')
  au <- paste(gsub('(AUTHOR)( IDENTIFIERS)*(: \\()','',au[1]), gsub('\\)','',au[2]),sep=', ')
  au <- toupper(au)

  # reference
  re <- gsub('.txt','',rev(strsplit(infile,'/')[[1]])[1])

  # read publications
  nskip <- ifelse (read.table(infile,sep=',',nrows=1,skip=1)=='null', 5, 4)
  dat <- read.table(infile, sep=',', stringsAsFactors=FALSE, skip=nskip)
  colnames(dat) <- read.table(infile,sep=',',stringsAsFactors=FALSE,skip=nskip-1,nrows=1)
  o <- rev(order(dat[,'Publication Year']))
  dat <- dat[o,]
  start <- min(dat[,'Publication Year'])
  w <- which(colnames(dat)==start)
  dat <- dat[,c(1:21,w:ncol(dat))]

  return(list(author=au,reference=re,pubs=dat))
}


R/citation_rank.R

#' Calculate citation ranking (quantiles)
#'
#' @param x A list with an element called  \code{pubs}, which is a
#' \code{data.frame} with columns \code{'Publication Year'} and
#' \code{'Total Citations'}.  Usually
#' from \code{\link{read.wos}}
#' @param w column indicators of which quantiles to grab
#' @param quant A \code{data.frame}? maybe
#'
#' @return A scalar number
#' @export
#'
#' @examples
#' wos_file = system.file("sbegueria.txt", package = "bibliometRics")
#' bib <- read.wos(wos_file)
#' hirsch(bib)
#' hirsch(bib, 2010)
#' egghe(bib)
#' egghe(bib, 2010)
#' base = read.baselines()
#' citation_rank(bib, quant=base$GEOSCIENCES)
citation_rank <- function(x, w = NULL, quant = NULL) {
  x <- x$pubs
  if (!is.null(w)) {
    x <- x[w, ]
  }
  r <-
    data.frame(a = x[, 'Publication Year'],
               b = x[, 'Total Citations'],
               c = '',
               stringsAsFactors = FALSE)
  r$c <-
    factor(r$c,
           levels = c(
             '>q0.9999',
             '>q0.999',
             '>q0.99',
             '>q0.9',
             '>q0.8',
             '>q0.5',
             '>q0'
           ))
  for (i in 1:nrow(r)) {
    w <- grep(r[i, 1], names(quant))
    if (length(w) == 0)
      next()
    ww <- which(r[i, 2] - quant[, w] >= 0)[1]
    if (is.na(ww)) {
      r[i, 3] <- levels(r$c)[7]
    } else {
      r[i, 3] <- levels(r$c)[ww]
    }
  }
  return(r[, 3])
}
#citation_rank(bib, c(9,10), quant)
#citation_rank(bib, quant)


R/bibliometRics.R




# read.scopus <- function(infile) {

	# # author
	# au <- read.table(infile,sep=',',nrows=1,stringsAsFactors=FALSE)[,2]
	# au <- substr(au,9,nchar(au))

	# # read publications
	# dat <- read.table(infile,sep=',',stringsAsFactors=FALSE,skip=7)[,c(1:7,9,12)]
	# colnames(dat) <- c('year','title','authors','issn','journal','volume','issue','cit','cittotal')

	# return()
# }
# bib <- read.scopus('scopus_SBP_win.csv')


#' Calculate Bibliometrics
#'
#' @param bib A list with an element called  \code{pubs}, which is a
#' \code{data.frame} with columns \code{'Publication Year'} and
#' \code{'Total Citations'}.  Usually
#' from \code{\link{read.wos}}
#' @param base Baseline percentages
#' @param discipline Name of Discipline, must be in column names of
#' \code{base}
#'
#' @return A \code{data.frame} of metric
#' @export
#'
#' @examples
#' wos_file = system.file("sbegueria.txt", package = "bibliometRics")
#' bib <- read.wos(wos_file)
#' bib$author = gsub('[ñÑ]', 'N', bib$author)
#' bib$pubs$Authors = gsub('[ñÑ]', 'N', bib$pubs$Authors)
#'
#' base = read.baselines()
#' discipline = 'GEOSCIENCES'
#' bibliometric(bib, base, 'GEOSCIENCES')
bibliometric <-
  function(bib,
           base = read.baselines(),
           discipline = NULL) {
    if (!is.null(discipline)) {
      w <- grep(discipline, names(base))
    } else {
      w <- 1
    }
    quant <- base[[w]]

    au <- bib$author
    dat <- bib$pubs

    # publications
    pubs <- nrow(dat)

    # publications as lead author
    leads <-
      as.character(lapply(dat[, 'Authors'], function(x) {
        strsplit(x, ';')[[1]][1]
      }))
    auu <- strsplit(au, ',')[[1]][1]
    pubs_lead <- grep(auu, leads, ignore.case = TRUE)
    #	pubs_lead <- length(grep(auu,leads,ignore.case=TRUE))

    # impact factor
    ifact2 <- round(ifactor(bib, n = 2), 2)
    ifact5 <- round(ifactor(bib, n = 5), 2)

    # citations
    cit_tot <- sum(dat[, 'Total Citations'])
    cit_max <- max(dat[, 'Total Citations'])
    cit_art <- round(mean(dat[, 'Total Citations']), 2)
    ini <- min(dat[, 'Publication Year'])
    span <- max(dat[, 'Publication Year']) - ini

    # number of pubs with n citations
    icit <- function(x, n = 10) {
      length(which(x[, 'Total Citations'] >= n))
    }
    i10 <- icit(dat)
    i25 <- icit(dat, 25)
    i50 <- icit(dat, 50)

    # h-index
    hin <- hirsch(bib)

    # g-index
    gin <- egghe(bib)

    # pubs >0.9, >0.99
    r <- citation_rank(bib, quant = quant)
    p09 <- sum(as.numeric(r) <= 4, na.rm = TRUE)
    p099 <- sum(as.numeric(r) <= 3, na.rm = TRUE)

    # pubs >0.9, as lead author
    r_lead <- citation_rank(bib, pubs_lead, quant)
    p09_lead <- sum(as.numeric(r_lead) <= 4, na.rm = TRUE)

    # i-score
    scores <-
      c(1 / (1 - 0.9999),
        1 / (1 - 0.999),
        1 / (1 - 0.99),
        1 / (1 - 0.9),
        1 / (1 - 0.8),
        1 / (1 - 0.5),
        1 / 1)
    iscore <- sum(scores[as.numeric(r)], na.rm = TRUE)

    # i-score, as lead author
    isc_lead <- sum(scores[as.numeric(r_lead)], na.rm = TRUE)

    # output
    out <- as.data.frame(t(
      c(
        name = au,
        ini,
        span,
        pubs,
        length(pubs_lead),
        round(pubs / span, 2),
        hin,
        round(hin / span, 2),
        gin,
        round(gin / span, 2),
        cit_tot,
        round(cit_tot / span, 2),
        cit_art,
        ifact2,
        ifact5,
        i10,
        i25,
        i50,
        cit_max,
        p09,
        p09_lead,
        p099,
        round(iscore, 2),
        round(isc_lead, 2)
      )
    ))
    colnames(out) <-
      c(
        'name',
        'ini',
        'years',
        'pubs',
        'lead',
        'pubs_year',
        'hin',
        'hin_year',
        'gin',
        'gin_year',
        'cit_tot',
        'cit_year',
        'cit_art',
        'ifact2',
        'ifact5',
        'i10',
        'i25',
        'i50',
        'cit_max',
        'pubs09',
        'pubs09_lead',
        'pubs099',
        'iscore',
        'iscore_lead'
      )
    return(out)
  }
#bibliometric(bib, baseline)








